include/protopuf/coder.h

```diff
@@ -17,42 +17,58 @@
 
 #include <utility>
 #include "byte.h"
+#include "coder_mode.h"
 
 namespace pp {
 
+    /// @brief A type which `encoder`'s `encode` returns.
+    /// @param Mode the encoding mode
+    template<coder_mode Mode>
+    using encode_result = typename Mode::template result_type<bytes>;
+
     /// @brief A pair type which `decoder`'s `decode` returns.
     /// - Left type of pair `T`: the type of decoded object.
     /// - Right type of pair `bytes`: the `bytes` which remains not decoded after finishing `decode`.
     template<typename T>
-    using decode_result = std::pair<T, bytes>;
+    using decode_value = std::pair<T, bytes>;
+
+    /// @brief A type which `decoder`'s `decode` returns.
+    /// @param T the type of decoded object
+    /// @param Mode the decoding mode
+    template<typename T, coder_mode Mode>
+    using decode_result = typename Mode::template result_type<decode_value<T>>;
 
     /// @brief Describes a type with static member function `encode`, which serializes an object to `bytes` (no ownership).
     ///
+    /// Encoding can be performed in different modes.
     /// Type alias `value_type` describes type of the object to be encoded.
     /// Static member function `encode`:
     /// @param v the object to be encoded (source object).
     /// @param s the bytes which the object `v` is encoded into (target bytes).
-    /// @returns a bytes from `begin(s) + encoding_length(v)` to `end(s)`, where `encoding_length` is the length of
+    /// @returns the @ref encode_result which depends on the encoding mode.
+    /// The result contains a bytes from `begin(s) + encoding_length(v)` to `end(s)`, where `encoding_length` is the length of
     /// encoded object (bytes form), representing the left bytes which remains not used yet.
     template<typename T>
     concept encoder = requires(typename T::value_type v, bytes s) {
         typename T::value_type;
-        { T::encode(v, s) } -> std::same_as<bytes>;
+        { T::template encode<unsafe_mode>(v, s) } -> std::same_as<encode_result<unsafe_mode>>;
```

Comment:
Maybe we can also add the `safe_mode` version to ensure that all coder implements the safe version `encode`/`decode`?

Reply:
Done